src/XtensaOptimize.cpp

```diff
@@ -2063,6 +2063,284 @@ class SimplifySliceConcat : public IRGraphMutator {
     SimplifySliceConcat() = default;
 };
 
+// Lifts Stack or VTCM allocation up the loop nest if possible.
+//
+// The idea is that we traverse a loop nest and look for allocations which are
+// of the right type (Stack or VTCM) and don't depend on any of the Let's
+// defined in the given loop. If these conditions are met then allocation can
+// be lifted out of the given loop (or potentially out of multiple of loops).
+// The difficulty is that due to loop partitioning there might be a multiple
+// instances of the same allocation in different instance of the loops. If all
+// instances can be lifted then we can have a single allocation outside of the
+// loop (we need to take max ot total extents as an extent for it to make sure
+// all instances are satisfied).
+// However, there might be a situation when only some of the allocations can be
+// lifted, which potentially could lead to duplicate allocation being generated.
+// Consider an example:
+// for yo:
+//   for xo:
+//        for x:
+//           allocate tmp[depends on let's inside of the loop x]
+//        for x:
+//           allocate tmp[doesn't depend on let's inside of the loop x]
+//        for x:
+//           allocate tmp[depends on let's inside of the loop x]
+// in which if we only consider a local context, the tmp allocation can be
+// lifted from the second inner loop, but couldn't be lifted from the first and
+// third loops.
+// In order to handle cases like this, LiftAllocations has two passes:
+// ** During the first pass, for each loop we compute a list of allocations which
+// are *not* allowed to be lifted to this loop level. This is done by finding a
+// *last* loop (i.e we are not able to lift allocation from the parent loop)
+// from which we *can* lift the allocation and then updating all of it's parent
+// loop lists to disallow given allocation.
+// ** In the second pass, we simply push allocations as far as we can up the
+// loop nest, while making sure that they satisfy all of the conditions (
+// including absence in the "disallowed" list).
+class LiftAllocations : public IRMutator {
+    using IRMutator::visit;
+
+    // Index of the current loop.
+    int loop_index = 0;
+    // Is this a first sub-pass of the pass?
+    bool first_run = true;
+
+    struct Allocation {
```

Comment:
Could uses of this struct be replaced with pointers to the original Allocate node?

Reply:
I did it for For node, but in this case it feels a bit more awkward because the original node is removed.